File: deep-learning/src/test/scala/org/apache/spark/dl/DeepVisionClassifier.scala
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in project root for information.

package org.apache.spark.dl

import com.microsoft.azure.synapse.ml.codegen.Wrappable
import com.microsoft.azure.synapse.ml.core.test.base.TestBase
import com.microsoft.azure.synapse.ml.logging.BasicLogging
import org.apache.spark.SparkConf
import org.apache.spark.ml.param.{Param, ParamMap, Params}
import org.apache.spark.ml.util.Identifiable
import org.apache.spark.ml.{ComplexParamsWritable, Estimator, Model}
import org.apache.spark.sql.functions.{col, udf}
import org.apache.spark.sql.types.{DoubleType, IntegerType, StructType}
import org.apache.spark.sql.{DataFrame, Dataset, Row, SparkSession}
import org.apache.spark.util.{Utils => SparkUtils}
import py4j.ClientServer

import java.io.File
import java.lang.ProcessBuilder.Redirect
import java.util


trait PredictionParams extends Params {
  val labelCol: Param[String] = new Param[String](this, "labelCol", "label column name")

  def setLabelCol(v: String): this.type

  def getLabelCol: String = $(labelCol)

  val imageCol: Param[String] = new Param[String](this, "imageCol", "image column name")

  def setImageCol(v: String): this.type

  def getImageCol: String = $(imageCol)

  val predictionCol: Param[String] = new Param[String](this, "predictionCol", "prediction column name")

  def setPredictionCol(v: String): this.type

  def getPredictionCol: String = $(predictionCol)

  setDefault(
    labelCol -> "label",
    imageCol -> "image",
    predictionCol -> "prediction"
  )

}

class DeepVisionClassifier(override val uid: String, val pythonEntryPoint: PythonEntryPoint)
  extends Estimator[DeepVisionModel] with PredictionParams
    with ComplexParamsWritable
    with Wrappable
    with BasicLogging {

  logClass()

  def this(pythonEntryPoint: PythonEntryPoint) =
    this(pythonEntryPoint.createObject("DeepVisionClassifier"), pythonEntryPoint)

  val backbone = new Param[String](this,
    "backbone",
    "Backbone of the deep vision model, should be a string representation of torchvision model.")

  def setBackbone(v: String): this.type = {
    val methodValue = new util.HashMap[String, Any]()
    methodValue.put("value", v)
    this.pythonEntryPoint.callMethod(this.uid, "setBackbone", methodValue)
    set(backbone, v)
  }

  def getBackbone(): String = $(backbone)

  val additionalLayersToTrain = new Param[Int](this,
    "additionalLayersToTrain",
    "number of last layers to fine tune for the model, should be between 0 and 3")

  def setAdditionalLayersToTrain(v: Int): this.type = {
    val methodValue = new util.HashMap[String, Any]()
    methodValue.put("value", v)
    this.pythonEntryPoint.callMethod(this.uid, "setAdditionalLayersToTrain", methodValue)
    set(additionalLayersToTrain, v)
  }

  def getAdditionalLayersToTrain(): Int = $(additionalLayersToTrain)

  val numClasses = new Param[Int](this,
    "numClasses",
    "number of target classes")

  def setNumClasses(v: Int): this.type = {
    val methodValue = new util.HashMap[String, Any]()
    methodValue.put("value", v)
    this.pythonEntryPoint.callMethod(this.uid, "setNumClasses", methodValue)
    set(numClasses, v)
  }

  def getNumClasses(): Int = $(numClasses)

  val lossName = new Param[String](this,
    "lossName",
    "string representation of torch.nn.functional loss function for the underlying " +
      "pytorch_lightning model, e.g. binary_cross_entropy")

  def setLossName(v: String): this.type = {
    val methodValue = new util.HashMap[String, Any]()
    methodValue.put("value", v)
    this.pythonEntryPoint.callMethod(this.uid, "setLossName", methodValue)
    set(lossName, v)
  }

  def getLossName(): String = $(lossName)

  val optimizerName = new Param[String](this,
    "optimizerName",
    "string representation of optimizer function for the underlying pytorch_lightning model")

  def setOptimizerName(v: String): this.type = {
    val methodValue = new util.HashMap[String, Any]()
    methodValue.put("value", v)
    this.pythonEntryPoint.callMethod(this.uid, "setOptimizerName", methodValue)
    set(optimizerName, v)
  }

  def getOptimizerName(): String = $(optimizerName)

  val dropoutAUX = new Param[Double](this,
    "dropoutAUX",
    "numeric value that's applied to googlenet InceptionAux module's dropout " +
      "layer only: probability of an element to be zeroed")

  def setDropoutAUX(v: Double): this.type = {
    val methodValue = new util.HashMap[String, Any]()
    methodValue.put("value", v)
    this.pythonEntryPoint.callMethod(this.uid, "setDropoutAUX", methodValue)
    set(dropoutAUX, v)
  }

  def getDropoutAUX(): Double = $(dropoutAUX)

  setDefault(additionalLayersToTrain -> 0,
    optimizerName -> "adam",
    lossName -> "cross_entropy",
    dropoutAUX -> 0.7)

  def setLabelCol(v: String): this.type = {
    val methodValue = new util.HashMap[String, Any]()
    methodValue.put("value", v)
    this.pythonEntryPoint.callMethod(this.uid, "setLabelCol", methodValue)
    set(labelCol, v)
  }

  def setImageCol(v: String): this.type = {
    val methodValue = new util.HashMap[String, Any]()
    methodValue.put("value", v)
    this.pythonEntryPoint.callMethod(this.uid, "setImageCol", methodValue)
    set(imageCol, v)
  }

  def setPredictionCol(v: String): this.type = {
    val methodValue = new util.HashMap[String, Any]()
    methodValue.put("value", v)
    this.pythonEntryPoint.callMethod(this.uid, "setPredictionCol", methodValue)
    set(predictionCol, v)
  }

  val storeParam = new Param[String](this,
    "storeParam",
    "A string representation of the store for horovod training")

  def setStoreParam(v: String): this.type = {
    val methodValue = new util.HashMap[String, Any]()
    methodValue.put("value", v)
    this.pythonEntryPoint.callMethod(this.uid, "setStoreParam", methodValue)
    set(storeParam, v)
  }

  def getStoreParam(): String = $(storeParam)

  val storePrefixPath = new Param[String](this,
    "storePrefixPath",
    "Prefix path used for store_param, should only be used when you use store_param instead of store")

  def setStorePrefixPath(v: String): this.type = {
    val methodValue = new util.HashMap[String, Any]()
    methodValue.put("value", v)
    this.pythonEntryPoint.callMethod(this.uid, "setStorePrefixPath", methodValue)
    set(storePrefixPath, v)
  }

  def getStorePrefixPath(): String = $(storePrefixPath)

  override def fit(dataset: Dataset[_]): DeepVisionModel = {
    logFit({
      val methodValue = new util.HashMap[String, Any]()
      methodValue.put("df", dataset)
      val result = this.pythonEntryPoint.callMethod(this.uid, "fit", methodValue)
        .asInstanceOf[Tuple2[String, Any]]
      val uid = result._1
      val deepVisionModelObj = result._2
      this.pythonEntryPoint.addObject(uid, deepVisionModelObj)
      new DeepVisionModel(uid, this.pythonEntryPoint)
    })
  }

  override def copy(extra: ParamMap): Estimator[DeepVisionModel] = defaultCopy(extra)

  override def transformSchema(schema: StructType): StructType = {
    schema.add(getPredictionCol, IntegerType)
  }
}


class DeepVisionModel(override val uid: String, val pythonEntryPoint: PythonEntryPoint)
  extends Model[DeepVisionModel] with PredictionParams
    with ComplexParamsWritable with BasicLogging {

  logClass()

  def this(pythonEntryPoint: PythonEntryPoint) =
    this(pythonEntryPoint.createObject("DeepVisionModel"), pythonEntryPoint)

  def setLabelCol(v: String): this.type = {
    val methodValue = new util.HashMap[String, Any]()
    methodValue.put("value", v)
    this.pythonEntryPoint.callMethod(this.uid, "setLabelCol", methodValue)
    set(labelCol, v)
  }

  def setImageCol(v: String): this.type = {
    val methodValue = new util.HashMap[String, Any]()
    methodValue.put("value", v)
    this.pythonEntryPoint.callMethod(this.uid, "setImageCol", methodValue)
    set(imageCol, v)
  }

  def setPredictionCol(v: String): this.type = {
    val methodValue = new util.HashMap[String, Any]()
    methodValue.put("value", v)
    this.pythonEntryPoint.callMethod(this.uid, "setPredictionCol", methodValue)
    set(predictionCol, v)
  }

  override def transform(dataset: Dataset[_]): DataFrame = {
    logTransform({
      val methodValue = new util.HashMap[String, Any]()
      methodValue.put("dataset", dataset)
      val df = this.pythonEntryPoint.callMethod(this.uid, "transform", methodValue)
      df.asInstanceOf[DataFrame]
    })
  }

  override def copy(extra: ParamMap): DeepVisionModel = defaultCopy(extra)

  override def transformSchema(schema: StructType): StructType = {
    schema.add(getPredictionCol, IntegerType)
  }

}

object Utils {
  def runCmd(cmd: Seq[String],
             wd: File = new File("."),
             envVars: Map[String, String] = Map()): Unit = {
    val pb = new ProcessBuilder()
      .directory(wd)
      .command(cmd: _*)
      .redirectError(Redirect.INHERIT)
      .redirectOutput(Redirect.INHERIT)
    val env = pb.environment()
    envVars.foreach(p => env.put(p._1, p._2))
    val result = pb.start().waitFor()
    if (result != 0) {
      println(s"Error: result code: ${result}")
      throw new Exception(s"Execution resulted in non-zero exit code: ${result}")
    }
  }

}

//object SparkObj {
//  var Spark: SparkSession = null
//
//  def init(): String = {
//    val sparkConfig = new SparkConf()
//    Spark = SparkSession.builder().config(sparkConfig)
//      .master("local[*]").appName("test py4j").getOrCreate()
//    val secret: String = SparkUtils.createSecret(sparkConfig)
//    secret
//  }
//
//  def shutdown(): Unit = Spark.stop()
//}

class TestSuite extends TestBase {
  val sparkConfig = spark.sparkContext.getConf
  val secret = SparkUtils.createSecret(sparkConfig)

  def generateData(): (DataFrame, DataFrame) = {
    import spark.implicits._

    val trainFolder = "/home/serena/repos/SynapseML/deep-learning/target/jpg/train"
    val testFolder = "/home/serena/repos/SynapseML/deep-learning/target/jpg/test"
    val trainFiles = new File(trainFolder).listFiles().filter(_.getName.endsWith(".jpg")).map(_.toString)
    val testFiles = new File(testFolder).listFiles().filter(_.getName.endsWith(".jpg")).map(_.toString)

    //    import spark.implicits._
    def extractLabel(path: String): Int = {
      path.split("/".toCharArray).takeRight(1)(0).split(".".toCharArray)
        .take(1)(0).split("_".toCharArray).takeRight(1)(0).toInt / 81
    }

    val extractLabelUdf = udf(extractLabel _)

    val trainDF = trainFiles.toSeq.toDF("image")
      .withColumn("label", extractLabelUdf(col("image")).cast(DoubleType))
    val testDF = testFiles.toSeq.toDF("image")
      .withColumn("label", extractLabelUdf(col("image")).cast(DoubleType))
    (trainDF, testDF)
  }

  // We should run python script first
  // This opens port 25334
  test("runPython") {
    Utils.runCmd(Seq("/home/serena/miniconda3/envs/synapseml/bin/python", "PythonEntryPoint.py", secret),
      new File("/home/serena/repos/SynapseML/deep-learning/src/test/scala/org/apache/spark/dl"))
  }

  test("deepVisionClassifier") {
    val clientServer = new ClientServer.ClientServerBuilder()
      .authToken("5abf6ac7b7dd62fd1878624894296a60af59642666c8d0198e67aaf026f8c064")
      .build()

    val pyEntryPoint = new PythonEntryPoint(clientServer)
    pyEntryPoint.updateGatewayImports()

    val deepVisionClassifier = new DeepVisionClassifier(pyEntryPoint)
      .setBackbone("resnet50")
      .setAdditionalLayersToTrain(1)
      .setNumClasses(17)
      .setStoreParam("LocalStore")
      .setStorePrefixPath("./test")

    println(deepVisionClassifier.getNumClasses())
    println(deepVisionClassifier.getLossName())
    println(deepVisionClassifier.getAdditionalLayersToTrain())

    val dfs = generateData()
    val trainDF = dfs._1
    val testDF = dfs._2

    val deepVisionModel = deepVisionClassifier.fit(trainDF)

    pyEntryPoint.shutdownPython()
    pyEntryPoint.shutdown()
  }

}

File: deep-learning/src/test/scala/org/apache/spark/dl/DeepVisionClassifier.py
# Copyright (C) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE in project root for information.

from lib2to3.pytree import convert
from multiprocessing.sharedctypes import Value
import sys

import torchvision.transforms as transforms
from horovod.spark.common.backend import SparkBackend
import horovod.spark.common.store as hstore
from horovod.spark.lightning import TorchEstimator
from PIL import Image
from pyspark.context import SparkContext
from pyspark.ml.param.shared import Param, Params
from lightning_utilities.core.imports import module_available
# from synapse.ml.dl.DeepVisionModel import DeepVisionModel
# from synapse.ml.dl.LitDeepVisionModel import LitDeepVisionModel
# from synapse.ml.dl.utils import keywords_catch
# from synapse.ml.dl.PredictionParams import PredictionParams
from PredictionParams import PredictionParams
from DeepVisionModel import DeepVisionModel
from LitDeepVisionModel import LitDeepVisionModel
from GatewayUtils import GatewayUtils
from utils import keywords_catch, generate_uuid
# from PythonEntryPoint import python_entry_point

from py4j.clientserver import ClientServer, JavaParameters, PythonParameters

_HOROVOD_AVAILABLE = module_available("horovod")
if _HOROVOD_AVAILABLE:
    import horovod

    _HOROVOD_EQUAL_0_25_0 = horovod.__version__ == "0.25.0"
    if not _HOROVOD_EQUAL_0_25_0:
        raise RuntimeError(
            "horovod should be of version 0.25.0, found: {}".format(horovod.__version__)
        )
else:
    raise ModuleNotFoundError("module not found: horovod")

class DeepVisionClassifier(TorchEstimator, PredictionParams):

    backbone = Param(
        Params._dummy(), "backbone", "backbone of the deep vision classifier"
    )

    additional_layers_to_train = Param(
        Params._dummy(),
        "additional_layers_to_train",
        "number of last layers to fine tune for the model, should be between 0 and 3",
    )

    num_classes = Param(Params._dummy(), "num_classes", "number of target classes")

    loss_name = Param(
        Params._dummy(),
        "loss_name",
        "string representation of torch.nn.functional loss function for the underlying pytorch_lightning model, e.g. binary_cross_entropy",
    )

    optimizer_name = Param(
        Params._dummy(),
        "optimizer_name",
        "string representation of optimizer function for the underlying pytorch_lightning model",
    )

    dropout_aux = Param(
        Params._dummy(),
        "dropout_aux",
        "numeric value that's applied to googlenet InceptionAux module's dropout layer only: probability of an element to be zeroed",
    )

    transform_fn = Param(
        Params._dummy(),
        "transform_fn",
        "A composition of transforms used to transform and augnment the input image, should be of type torchvision.transforms.Compose",
    )

    store_param = Param(
        Params._dummy(),
        "store_param",
        "A string representation of the store for horovod training",
    )

    store_prefix_path = Param(
        Params._dummy(),
        "store_prefix_path",
        "Prefix path used for store_param, should only be used when you use store_param instead of store",
    )

    @keywords_catch
    def __init__(
            self,
            backbone=None,
            additional_layers_to_train=0,
            num_classes=None,
            optimizer_name="adam",
            loss_name="cross_entropy",
            dropout_aux=0.7,
            transform_fn=None,
            # Classifier args
            label_col="label",
            image_col="image",
            prediction_col="prediction",
            # TorchEstimator args
            num_proc=None,
            backend=None,
            store=None,
            metrics=None,
            loss_weights=None,
            sample_weight_col=None,
            gradient_compression=None,
            input_shapes=None,
            validation=None,
            callbacks=None,
            batch_size=None,
            val_batch_size=None,
            epochs=None,
            verbose=1,
            random_seed=None,
            shuffle_buffer_size=None,
            partitions_per_process=None,
            run_id=None,
            train_minibatch_fn=None,
            train_steps_per_epoch=None,
            validation_steps_per_epoch=None,
            transformation_fn=None,
            train_reader_num_workers=None,
            trainer_args=None,
            val_reader_num_workers=None,
            reader_pool_type=None,
            label_shapes=None,
            inmemory_cache_all=False,
            num_gpus=None,
            logger=None,
            log_every_n_steps=50,
            data_module=None,
            loader_num_epochs=None,
            terminate_on_nan=False,
            profiler=None,
            debug_data_loader=False,
            train_async_data_loader_queue_size=None,
            val_async_data_loader_queue_size=None,
            use_gpu=True,
            mp_start_method=None,
    ):
        super(DeepVisionClassifier, self).__init__()

        self._setDefault(
            backbone=None,
            additional_layers_to_train=0,
            num_classes=None,
            optimizer_name="adam",
            loss_name="cross_entropy",
            dropout_aux=0.7,
            transform_fn=None,
            feature_cols=["image"],
            label_cols=["label"],
            label_col="label",
            image_col="image",
            prediction_col="prediction",
        )

        kwargs = self._kwargs
        self._set(**kwargs)

    def setBackbone(self, value):
        return self._set(backbone=value)

    def getBackbone(self):
        return self.getOrDefault(self.backbone)

    def setAdditionalLayersToTrain(self, value):
        return self._set(additional_layers_to_train=value)

    def getAdditionalLayersToTrain(self):
        return self.getOrDefault(self.additional_layers_to_train)

    def setNumClasses(self, value):
        return self._set(num_classes=value)

    def getNumClasses(self):
        return self.getOrDefault(self.num_classes)

    def setLossName(self, value):
        return self._set(loss_name=value)

    def getLossName(self):
        return self.getOrDefault(self.loss_name)

    def setOptimizerName(self, value):
        return self._set(optimizer_name=value)

    def getOptimizerName(self):
        return self.getOrDefault(self.optimizer_name)

    def setDropoutAUX(self, value):
        return self._set(dropout_aux=value)

    def getDropoutAUX(self):
        return self.getOrDefault(self.dropout_aux)

    def setTransformFn(self, value):
        return self._set(transform_fn=value)

    def getTransformFn(self):
        return self.getOrDefault(self.transform_fn)
    
    def setStoreParam(self, value):
        if self.getStore() is not None:
            raise ValueError("Only one of store and storeParam can be set, storeParam is the string representation of store")
        valid_stores = ("LocalStore", "HDFSStore", "DBFSLocalStore")
        if value not in valid_stores:
            raise ValueError(f"StoreParam should be one of {valid_stores}, received: {value}")
        return self._set(store_param=value)
    
    def getStoreParam(self):
        return self.getOrDefault(self.store_param)
    
    def setStorePrefixPath(self, value):
        if self.getStore() is not None:
            raise ValueError("No need to set storePrefixPath if you use store instead of storeParam")
        return self._set(store_prefix_path=value)
    
    def getStorePrefixPath(self):
        return self.getOrDefault(self.store_prefix_path)

    def _update_input_shapes(self):
        if self.getInputShapes() is None:
            if self.getBackbone().startswith("inception"):
                self.setInputShapes([[-1, 3, 299, 299]])
            else:
                self.setInputShapes([[-1, 3, 224, 224]])

    def _update_cols(self):
        self.setFeatureCols([self.getImageCol()])
        self.setLabelCols([self.getLabelCol()])
    
    def _update_store(self):
        if self.getStore() is None:
            if self.getStoreParam() is None:
                raise ValueError(f"At least one of store and storeParam needs to be set")
            if self.getStorePrefixPath() is None:
                raise ValueError(f"PrefixPath of {self.getStoreParam()} is not set")
            constructed_store = hstore.__dict__[self.getStoreParam()](prefix_path=self.getStorePrefixPath())
            self.setStore(constructed_store)
    
    def _fit_on_parquet(self, dataset_idx=None):
        self._update_store()
        return super()._fit_on_parquet(dataset_idx)

    def _fit(self, dataset):
        from pyspark.ml.common import _py2java, _java2py
        from pyspark.sql.session import SparkSession
        import logging

        sparkContext = self._get_active_spark_context()

        logging.warning(f"Active Spark Session: {SparkSession.getActiveSession()}")
        dataset = _java2py(sparkContext, dataset)
        logging.warning(f"dataset type: {type(dataset)}")

        self._update_input_shapes()
        self._update_cols()
        self._update_transformation_fn()
        self._update_store()
        
        model = LitDeepVisionModel(
            backbone=self.getBackbone(),
            additional_layers_to_train=self.getAdditionalLayersToTrain(),
            num_classes=self.getNumClasses(),
            input_shape=self.getInputShapes()[0],
            optimizer_name=self.getOptimizerName(),
            loss_name=self.getLossName(),
            label_col=self.getLabelCol(),
            image_col=self.getImageCol(),
            dropout_aux=self.getDropoutAUX(),
        )
        self._set(model=model)

        logging.warning(f"model created: {model}")
        deepVisionModel = super()._fit(dataset)
        logging.warning(f"-------fit completion: {deepVisionModel}")
        # Set python_entry_point to global variable
        # global python_entry_point
        uid = generate_uuid("DeepVisionModel")
        # python_entry_point.addObject(uid, deepVisionModel)
        return uid, deepVisionModel

    # override this method to provide a correct default backend
    def _get_or_create_backend(self):
        backend = self.getBackend()
        num_proc = self.getNumProc()
        if backend is None:
            if num_proc is None:
                num_proc = self._find_num_proc()
            backend = SparkBackend(
                num_proc,
                stdout=sys.stdout,
                stderr=sys.stderr,
                prefix_output_with_timestamp=True,
                verbose=self.getVerbose(),
            )
        elif num_proc is not None:
            raise ValueError(
                'At most one of parameters "backend" and "num_proc" may be specified'
            )
        return backend

    def _find_num_proc(self):
        if self.getUseGpu():
            # set it as number of executors for now (ignoring num_gpus per executor)
            sc = SparkContext.getOrCreate()
            return sc._jsc.sc().getExecutorMemoryStatus().size() - 1
        return None

    def _update_transformation_fn(self):
        if self.getTransformationFn() is None:
            if self.getTransformFn() is None:
                crop_size = self.getInputShapes()[0][-1]
                transform = transforms.Compose(
                    [
                        transforms.RandomResizedCrop(crop_size),
                        transforms.RandomHorizontalFlip(),
                        transforms.ToTensor(),
                        transforms.Normalize(
                            mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]
                        ),
                    ]
                )
                self.setTransformFn(transform)

            def _create_transform_row(image_col, label_col, transform):
                def _transform_row(row):
                    path = row[image_col]
                    label = row[label_col]
                    image = Image.open(path).convert("RGB")
                    image = transform(image).numpy()
                    return {image_col: image, label_col: label}

                return _transform_row

            self.setTransformationFn(
                _create_transform_row(
                    self.getImageCol(),
                    self.getLabelCol(),
                    self.getTransformFn(),
                )
            )

    def get_model_class(self):
        return DeepVisionModel

    def _get_model_kwargs(self, model, history, optimizer, run_id, metadata):
        return dict(
            history=history,
            model=model,
            optimizer=optimizer,
            input_shapes=self.getInputShapes(),
            run_id=run_id,
            _metadata=metadata,
            loss=self.getLoss(),
            loss_constructors=self.getLossConstructors(),
            label_col=self.getLabelCol(),
            image_col=self.getImageCol(),
            prediction_col=self.getPredictionCol(),
        )
    
    # For py4j only
    def _get_object_id(self):
        return self.uid
